Add keyboard padding and margin feature

percula opened the pull request:
Typically, we use "adjustResize" in the manifest to automatically resize an activity so that the keyboard doesn't cover up the content. However, edge-to-edge breaks that. This PR adds the ability to add padding or margin, equivalent to the height of the keyboard (same as `SystemWindowInsetBottom`), to the bottom of a view when the keyboard is shown.

Included in this PR:

- Java Insetter method
- Extension function
- Databinding adapter
- Addition to DataBindingSample sample to test that it works

chrisbanes commented on dbx/src/main/java/dev/chrisbanes/insetter/InsetterBindingAdapters.java:
```diff
@@ -102,6 +102,22 @@ public void onApplyInsets(
 
+  @BindingAdapter(
+            value = {
+                    "marginBottomForKeyboard",
```
Can we rename this to `marginBottomForIme`? Since this is only needed for on screen keyboards, I think it's better to be clear

chrisbanes commented on dbx/src/main/java/dev/chrisbanes/insetter/InsetterBindingAdapters.java:
```diff
@@ -102,6 +102,22 @@ public void onApplyInsets(
+  @BindingAdapter(
+            value = {
+                    "marginBottomForKeyboard",
+                    "paddingBottomForKeyboard"
```
Ditto, `...forIme`

chrisbanes commented on ktx/src/main/java/dev/chrisbanes/insetter/viewinsetter.kt:
```diff
@@ -45,4 +45,14 @@ fun View.requestApplyInsetsWhenAttached() {
+ * @param marginBottom whether to adjust the view's bottom margin
+ * @param paddingBottom whether to adjust the view's bottom padding
+ */
+fun View.applyInsetsWhenKeyboardDisplayed(marginBottom: Boolean = false,
```
Ime

chrisbanes commented on library/src/main/java/dev/chrisbanes/insetter/Insetter.java:
```diff
@@ -332,6 +334,89 @@ public static void setEdgeToEdgeSystemUiFlags(@NonNull final View view, final bo
+    view.getWindowVisibleDisplayFrame(visibleWindowBounds);
+    int visibleWindowHeight = visibleWindowBounds.height();
+    int heightDiff = view.getRootView().getHeight() - visibleWindowHeight;
+    int marginOfError = Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 100, view.getResources().getDisplayMetrics()));
```
Can we move this is a dimension resource? Easier to maintain and we can vary on API level/resource qualifiers if needed

chrisbanes commented on library/src/main/java/dev/chrisbanes/insetter/Insetter.java:
```diff
@@ -332,6 +334,89 @@ public static void setEdgeToEdgeSystemUiFlags(@NonNull final View view, final bo
+   * Determines if the keyboard is open by comparing the visible window height to the height
+   */
+  private static boolean isKeyboardOpen(View view) {
+    Rect visibleWindowBounds = new Rect();
```
Nit: cache this in a `ThreadLocal`

chrisbanes commented:
...oh and a test if possible please!

percula commented:
You're welcome, happy to help! Thanks for this wonderful library and your excellent blog posts! I implemented your comments (although I might have misinterpreted your comment about `ThreadLocal`). Also, no test at the moment, sorry.

So I discovered an issue when the keyboard opens. Hard to explain so here's some gifs. I added a icon where a FAB would usually go to illustrate the issue:

Case 1, without any of my changes. In this case, the `systemWindowInsetBottom` increases to include the navbar and the keyboard, bumping the FAB up. However, the EditText becomes hidden behind the keyboard.
![BeforePR44](https://user-images.githubusercontent.com/16809252/74155302-5d67d780-4be2-11ea-9f0b-3804d6ed0064.gif)

Case 2, with my initial changes. In this case, the `systemWindowInsetBottom` increases to include the nav bar and the keyboard bumping the FAB up. In addition, the FrameLayout's bottom padding increases to show the full EditText. This results in the FAB being way out-of-place.
![WithoutFix](https://user-images.githubusercontent.com/16809252/74155688-ff87bf80-4be2-11ea-982d-396012c0ac76.gif)

Case 3, with the fix for this. In this case, the bottom margin of the FAB only includes `systemWindowInsetBottom` when the keyboard is not shown.
![WithFix](https://user-images.githubusercontent.com/16809252/74155869-44abf180-4be3-11ea-924a-afcae7445ec9.gif)

I think this is a good solution, since when the keyboard is shown we really don't need any extra padding or margin due to the bottom inset.

chrisbanes commented on ktx/src/main/java/dev/chrisbanes/insetter/viewinsetter.kt:
```diff
@@ -45,4 +45,14 @@ fun View.requestApplyInsetsWhenAttached() {
+ * @param marginBottom whether to adjust the view's bottom margin
+ * @param paddingBottom whether to adjust the view's bottom padding
+ */
+fun View.applyInsetsWhenIMEDisplayed(marginBottom: Boolean = false,
```
`applyInsetsWhenImeIsDisplayed`

chrisbanes commented on library/src/main/java/dev/chrisbanes/insetter/Insetter.java:
```diff
@@ -141,6 +142,7 @@ public static void applyInsetsToView(
               insets,
               initialState));
     }
+    Boolean isIMEOpen = Insetter.isIMEOpen.update(view).get();
```
`boolean` ?

chrisbanes commented on library/src/main/java/dev/chrisbanes/insetter/Insetter.java:
```diff
@@ -332,6 +342,96 @@ public static void setEdgeToEdgeSystemUiFlags(@NonNull final View view, final bo
+    setOnApplyInsetsListener(v, new OnApplyInsetsListener() {
+      @Override
+      public void onApplyInsets(@NonNull View view, @NonNull WindowInsetsCompat insets, @NonNull ViewState initialState) {
+        Boolean isIMEOpen = Insetter.isIMEOpen.update(view).get();
```
`boolean`

chrisbanes commented on library/src/main/java/dev/chrisbanes/insetter/Insetter.java:
```diff
@@ -332,6 +342,96 @@ public static void setEdgeToEdgeSystemUiFlags(@NonNull final View view, final bo
+   */
+  private static IMEOpen isIMEOpen = new IMEOpen();
+
+  private static class IMEOpen extends ThreadLocal<Boolean> {
```
Woah, this isn't what I meant. My bad for now explaining. I meant to cache the rect in a threadlocal so we don't keep creating them. Forget that comment for now and revert back to the previous code :smile:

chrisbanes commented:
Superceded by #88
